pkg/server/sotw/v3/server.go

```diff
@@ -255,3 +236,27 @@ func (s *server) StreamHandler(stream stream.Stream, typeURL string) error {
 
 	return s.process(stream, reqCh, typeURL)
 }
+
+func (s *server) registerResourceNames(resources []string, streamState *stream.StreamState) {
+	if len(resources) == 0 && streamState.IsWildcard() {
+		// The xDS protocol states that if there has never been any resource set, the request should be considered wildcard
+		// This would theoretically require keeping track on whether we ever became non-empty.
+		// As it is also technically allowed to return resources which have not been subscribed to, it is a best effort here.
```

Comment:
I don't really understand this comment... Is the scenario you're thinking of:

```
- Stream requests some resources (not considered wildcard)?
- Then unsubscribes those resources, hence ending up with no resources
In theory this should not be considered wildcard, but for us it does and it's fine since we only end up sending too unrequested resources?
```

Reply:
The case is: 
- stream is not wildcard
- then is made wildcard
- then has no resource at all
According to xDS this should be considered as "no resources", but this requires keeping a history of whether we ever left legacy mode (i.e. did we ever have resources set in any way, including *)
As the protocol also states that it's okay to return too much, I chose to ignore this case (which should anyway basically never occur in envoy today)